src/System.Private.CoreLib/shared/System/Object.cs

```diff
@@ -44,12 +45,12 @@ public virtual string ToString()
         // Equal to this.  Equality is defined as object equality for reference
         // types and bitwise equality for value types using a loader trick to
         // replace Equals with EqualsValue for value types).
-        public virtual bool Equals(object obj)
+        public virtual bool Equals(object? obj)
```

Comment:
Should ToString return `string?`?  This gets back to the question of how we want to handle base virtuals.  I guarantee you that there are ToString overrides that return null (I'm pretty sure we even have a few in coreclr/corefx), even though design guidelines say you shouldn't do that.

Reply:
For other types we haven’t annotated the return type of ToString as nullable. However this depend on what we want to do. I guess if the design guidelines say you shouldn’t return null wouldn’t this be a good reason to leave it as non-nullable?

Reply:
> I guess if the design guidelines say you shouldn’t return null wouldn’t this be a good reason to leave it as non-nullable?

Except then if something did return null, a caller that does ToString().Length won't get a warning and will null ref.

Reply:
> For other types we haven’t annotated the return type of ToString as nullable.

@jaredpar can correct me if I'm wrong, but I believe we allow for overrides to be more constrained than a base, e.g. an override could be `string` with the base being `string?`.

Reply:
I'll wait for @jaredpar  to confirm, if overrides can be more contrained, then it makes sense to make `Object.ToString()` return `string?`

Reply:
@gafter, @jcouv as well.

Yes overrides can be more strict than base calls.

Reply:
Yes.  Unfortunately, when you type "override ToString" and let the IDE complete writing the method, you'll get a nullable return type.

Reply:
Hmm. Do believe returning null is common or desirable? If we believe it's not I think we should mark it non-nullable.

Reply:
> Yes overrides can be more strict than base calls.

I just tried to mark `Object.ToString()` to return `string?` and then I get an error on `String.ToString()` which doesn't return nullable.

```
shared\System\String.cs(531,32): error CS8609: Nullability of reference types in return type doesn't match overridden member. [E:\repos\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
```

Reply:
@safern, we need to upgrade our compiler bits. It works int he current master: https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgiBiAOwFcAbKiYKuDOMuhoolAZiwCYMBhDAG8iGMVm4A3AJYIYFCFSzY0AfgwBZABQBKYaPGGUAdgyUaBQoYC+RW8UJdeGACIYQA/VfFOA9pMQEaQATRhxMbT0Rb0MxEwwAIgTLGzsgA===

Reply:
Actually, looks like it works and gives a warning matching what you gave above.

Possibly a `/warnaserror` scenario?

Reply:
It also looks to work in either direction and gives the same warning in either case.

@jaredpar, is that expected?

Reply:
@gafter

Reply:
I would have expected that, if not matching was allowed, then narrowing (going from nullable to non-nullable) works, without a warning and widening (going from non-nullable to nullable) doesn't work (because virtuals and that means the base could be "lying").
* This is, of course, assuming both types in the hierarchy have opted-into nullable, since things all go out the window otherwise

Reply:
The intent is to allow for parameters to allow widening on overrides, and narrowing on returns. There is some work in the compiler left to implement this completely but I thought that was in the `interface` case only.

Reply:
For reference, here is sharplab showing both cases using last night's master bits:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgiBiAOwFcAbKiYKuDOMuhoolAZiwCYMBhDAG8iGMVm4A3AJYIYFCFSzY0AfgwBZABQBKYaPGGUAdgyUaBQoYC+BsXYkYZchUpyYAcrv1XD4kxgARIGWNkS2xIRcvBgAIhggAj5G3AD2kogI0gAmjO6a3iK+flimwaHiEYYO0emZOXkq6l56RSViAeZUFWIR1kA=

Reply:
@tannergooding We've not implemented the narrowing/widening rules in overrides yet.
Tracked by https://github.com/dotnet/roslyn/issues/23268